src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/validation/OcrValidator.java

```diff
@@ -0,0 +1,105 @@
+package uk.gov.hmcts.reform.bulkscanprocessor.validation;
+
+import io.vavr.control.Try;
+import joptsimple.internal.Strings;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.data.util.Optionals;
+import org.springframework.stereotype.Component;
+import uk.gov.hmcts.reform.authorisation.generators.AuthTokenGenerator;
+import uk.gov.hmcts.reform.bulkscanprocessor.config.ContainerMappings;
+import uk.gov.hmcts.reform.bulkscanprocessor.exceptions.OcrValidationException;
+import uk.gov.hmcts.reform.bulkscanprocessor.model.blob.InputEnvelope;
+import uk.gov.hmcts.reform.bulkscanprocessor.model.blob.InputScannableItem;
+import uk.gov.hmcts.reform.bulkscanprocessor.ocrvalidation.client.OcrValidationClient;
+import uk.gov.hmcts.reform.bulkscanprocessor.ocrvalidation.client.model.req.FormData;
+import uk.gov.hmcts.reform.bulkscanprocessor.ocrvalidation.client.model.req.OcrDataField;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+import static java.util.stream.Collectors.toList;
+
+@Component
+@EnableConfigurationProperties(ContainerMappings.class)
+public class OcrValidator {
+
+    private static final Logger log = LoggerFactory.getLogger(OcrValidator.class);
+
+    private final OcrValidationClient client;
+    private final ContainerMappings containerMappings;
+    private final AuthTokenGenerator authTokenGenerator;
+
+    //region constructor
+    public OcrValidator(
+        OcrValidationClient client,
+        ContainerMappings containerMappings,
+        AuthTokenGenerator authTokenGenerator
+    ) {
+        this.client = client;
+        this.containerMappings = containerMappings;
+        this.authTokenGenerator = authTokenGenerator;
+    }
+    //endregion
+
+    public void assertIsValid(InputEnvelope envelope) {
+        Optionals.ifAllPresent(
+            findValidationUrl(envelope),
+            findDocWithOcr(envelope),
+            (validationUrl, docWithOcr) ->
+                Try.of(() -> client.validate(validationUrl, toFormData(docWithOcr), authTokenGenerator.generate()))
+                    .onSuccess(res -> {
+                        switch (res.status) {
+                            case ERRORS:
+                                throw new OcrValidationException(Strings.join(res.errors, ", "));
+                            case WARNINGS:
+                                log.info("Validation ended with warnings. File name: {}", envelope.zipFileName);
+                                break;
+                            default:
+                                break;
+                        }
+                    })
+                    .onFailure(exc -> {
+                        log.error("Error calling validation endpoint. Url: {}", validationUrl, exc);
+                        // log error and proceed
+                    })
```

Comment:
would be nice to have a biconsumer reference used here giving the current signature of the method. would be even nicer if, somehow, is possible to reduce usage of `option` and `try` in one place as separate core elements here

Reply:
Em this now definitely looks big enough - candidate for a split. Whichever way, biconsumer, onsuccess/onfailure registrar, both 🙃 🤷‍♂